### PR TITLE
Add openssl support in transport module with some enhancement and bug fix

### PR DESCRIPTION
**Pull request by ostarJoy**
This pr contain three things:
1. Fix bug in `AutoEvidence` which cause incomplete TEE detect.
2. Add tls support using openssl in transport module.
3. Create x509 certificate manually to pass openssl cert pre-verification.
4. Add tls echo client and echo server exmpale.

### CODE REVIEW

File: rats-rs/src/transport/tls/mod.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            SslInit::LOAD_SSL_STRINGS.bits() | SslInit::LOAD_CRYPTO_STRINGS.bits(),
+            ptr::null(),
+        );
+        OPENSSL_init_crypto(SslInit::LOAD_CRYPTO_STRINGS.bits(), ptr::null());
+        OPENSSL_init_crypto(SslInit::ADD_ALL_DIGESTS.bits(), ptr::null());
+    });
+    if unsafe { OPENSSL_init_ssl(0, ptr::null()) } < 0 {
+        return Err(Error::kind(ErrorKind::OsslInitializeFail));
+    }
+    Ok(())
```
Comment (imlk0) [inline rats-rs/src/transport/tls/mod.rs:106]: As mentioned at the OpenSSL [doc](https://docs.openssl.org/1.1.1/man3/OPENSSL_init_ssl/):
```txt
As of version 1.1.0 OpenSSL will automatically allocate all resources that it needs so no explicit initialisation is required.
```

Could you explain that why we need a explicit initialisation for OpenSSL here?

File: rats-rs/src/transport/tls/client.rs [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+        }
+    }
+}
+
+#[maybe_async]
+impl GenericSecureTransPortWrite for Client {
+    async fn send(&mut self, bytes: &[u8]) -> Result<()> {
+        if self.ctx.is_none() || self.ssl_session.is_none() {
+            return Err(Error::kind(ErrorKind::OsslCtxOrSessionUninitialized));
+        }
```
Comment (imlk0) [inline rats-rs/src/transport/tls/client.rs]: It looks that `Client::init()` is called by `TlsClientBuilder::build()` only. What about moving the code of `Client::init()` into `TlsClientBuilder::build()`? In this way, we can remove the `Option<>` from the types of `ctx` and `ssl_session`, thus avoiding the `is_none()` check here.

File: rats-rs/src/transport/tls/server.rs [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+        self
+    }
+}
+
+#[maybe_async]
+impl GenericSecureTransPortWrite for Server {
+    async fn send(&mut self, bytes: &[u8]) -> Result<()> {
+        if self.ctx.is_none() || self.ssl_session.is_none() {
+            return Err(Error::kind(ErrorKind::OsslCtxOrSessionUninitialized));
+        }
```
Comment (imlk0) [inline rats-rs/src/transport/tls/server.rs]: Also check here, as with `struct Client`.

File: rats-rs/src/transport/tls/mod.rs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        Arc::new(Mutex::new(Cell::new(CRYPTO_get_ex_new_index(
+            4,
+            0,
+            ptr::null_mut(),
+            None,
+            None,
+            None,
+        ))))
+    };
+}
```
Comment (imlk0) [inline rats-rs/src/transport/tls/mod.rs]: Can be replaced with:

```suggestion
lazy_static! {
    static ref OPENSSL_EX_DATA_IDX: i32 =
        unsafe { CRYPTO_get_ex_new_index(4, 0, ptr::null_mut(), None, None, None,) };
}
```

File: rats-rs/src/transport/tls/server.rs [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+            }
+        }
+        let session = unsafe { SSL_new(ctx) };
+        if session.is_null() {
+            return Err(Error::kind(ErrorKind::OsslNoMem));
+        }
+        unsafe {
+            X509_STORE_set_ex_data(
+                SSL_CTX_get_cert_store(ctx),
+                OPENSSL_EX_DATA_IDX.lock().unwrap().get(),
```
Comment (imlk0) [inline rats-rs/src/transport/tls/server.rs]: Can be replaced with:

```suggestion
                *OPENSSL_EX_DATA_IDX,
```

File: rats-rs/src/transport/tls/client.rs [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+            }
+        }
+        let session = unsafe { SSL_new(ctx) };
+        if session.is_null() {
+            return Err(Error::kind(ErrorKind::OsslNoMem));
+        }
+        unsafe {
+            X509_STORE_set_ex_data(
+                SSL_CTX_get_cert_store(ctx),
+                OPENSSL_EX_DATA_IDX.lock().unwrap().get(),
```
Comment (imlk0) [inline rats-rs/src/transport/tls/client.rs]: Can be replaced with:

```suggestion
                *OPENSSL_EX_DATA_IDX,
```

File: rats-rs/src/transport/tls/mod.rs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    fn get_fd(&self) -> i32;
+}
+
+struct GetFdDumpImpl;
+
+impl GetFd for GetFdDumpImpl {
+    fn get_fd(&self) -> i32 {
+        0
+    }
+}
```
Comment (imlk0) [inline rats-rs/src/transport/tls/mod.rs]: Could you clarify what the purpose of GetFdDumpImpl is?

**Review by imlk0 — CHANGES_REQUESTED** [review]
A great work. And I've left some comments, pls check them :)

**Comment by ostarJoy** [comment]
> A great work. And I've left some comments, pls check them :)

Thank you for your suggestion. I have made revisions to the issues raised in last commit :)

**Pull request merged**